### ISSUE TITLE
Set parameters to obtain same output between eqx.nn.ConvTranspose2d and torch.nn.ConvTranspose2d

### ISSUE THREAD

**Issue by chMoussa** [patrick-kidger/equinox#574]
Hey, I am trying to set the weights and biases between instances of eqx.nn.ConvTranspose2d and torch.nn.ConvTranspose2d. 
However, I am unsuccessful in doing so, as the weights have different shape ordering between both implementations. I saw a [similar issue with conv2d](https://github.com/patrick-kidger/equinox/issues/171). Below is an example code replicating the problem. How to fix this?

`import jax
import jax.numpy as jnp
import jax.tree_util as jtu

import equinox as eqx
import torch.nn as tnn
import torch
import numpy as np
latent_space_dim = 100
hidden_features = 64
np.random.seed(0)
z = np.random.randn(100, 100, 1, 1).astype(np.float64)
key = jax.random.PRNGKey(0)
torch.manual_seed(0)
t_cv = tnn.ConvTranspose2d(
                latent_space_dim,
                hidden_features * 8,
                kernel_size=4,
                stride=1,
                padding=0,
            )

e_cv = eqx.nn.ConvTranspose2d(
                latent_space_dim,
                hidden_features * 8,
                kernel_size=4,
                stride=1,
                padding=0,
                key=key,
            )

t_w, t_b = t_cv.weight.detach().numpy().reshape(e_cv.weight.shape), t_cv.bias.detach().numpy().reshape(e_cv.bias.shape)
_, treedef = jtu.tree_flatten(e_cv)
e_cv = jtu.tree_unflatten(treedef, [jnp.asarray(t_w), jnp.asarray(t_b)])

class ConvGenerator(eqx.Module):
    layers: list

    def __init__(
        self,
    ) -> None:
        if not isinstance(hidden_features, int):
            raise TypeError("hidden_features must be an integer")

        self.layers = [
            e_cv
        ]

    def __call__(self, z):
        for l in self.layers:
            z = l(z)
        return z

gen = tnn.Sequential(t_cv)
gen_jax = ConvGenerator()

np.random.seed(0)
z = np.random.randn(100, 100, 1, 1)
z_jnp = jnp.array(z)
z_torch = torch.Tensor(z).detach()
output_gen_torch = gen(z_torch).detach()
output_gen_jax = eqx.filter_vmap(gen_jax)(z_jnp)

assert jnp.isclose(output_gen_torch.numpy(), output_gen_jax).all()

`

**Comment by Artur-Galstyan** [comment]
As a little help, here's the code formatted and simplified:

```python
import equinox as eqx
import jax
import jax.numpy as jnp
import jax.tree_util as jtu
import numpy as np
import torch
import torch.nn as tnn
from icecream import ic

latent_space_dim = 2
hidden_features = 1
np.random.seed(0)
z = np.random.randn(latent_space_dim, latent_space_dim, 1, 1).astype(np.float64)
key = jax.random.PRNGKey(0)
torch.manual_seed(0)

t_cv = tnn.ConvTranspose2d(
    latent_space_dim,
    hidden_features * 8,
    kernel_size=4,
    stride=1,
    padding=0,
)

ic(t_cv.weight.shape)
ic(t_cv.bias.shape)

e_cv = eqx.nn.ConvTranspose2d(
    latent_space_dim,
    hidden_features * 8,
    kernel_size=4,
    stride=1,
    padding=0,
    key=key,
)

ic("before", e_cv.weight.shape)
ic("before", e_cv.bias.shape)

t_w, t_b = (
    t_cv.weight.detach().numpy().reshape(e_cv.weight.shape),
    t_cv.bias.detach().numpy().reshape(e_cv.bias.shape),
)
_, treedef = jtu.tree_flatten(e_cv)
e_cv = jtu.tree_unflatten(treedef, [jnp.asarray(t_w), jnp.asarray(t_b)])

# other way of model surgery
# new_weight = t_w
# where = lambda l: l.weight
# e_cv = eqx.tree_at(where, e_cv, new_weight)

# new_bias = t_b
# where = lambda l: l.bias
# e_cv = eqx.tree_at(where, e_cv, new_bias)

ic("after", e_cv.weight.shape)
ic("after", e_cv.bias.shape)


gen = tnn.Sequential(t_cv)
# gen_jax = ConvGenerator()
gen_jax = e_cv

np.random.seed(0)
z = np.random.randn(latent_space_dim, latent_space_dim, 1, 1)
z_jnp = jnp.array(z)
z_torch = torch.Tensor(z).detach()
output_gen_torch = gen(z_torch).detach()
output_gen_jax = eqx.filter_vmap(gen_jax)(z_jnp)

ic(output_gen_torch.shape)
ic(output_gen_jax.shape)


assert jnp.isclose(output_gen_torch.numpy(), output_gen_jax).all()
```

I see that the output is 
```
ic| t_cv.weight.shape: torch.Size([2, 8, 4, 4])
ic| t_cv.bias.shape: torch.Size([8])
ic| 'before', e_cv.weight.shape: (8, 2, 4, 4)
ic| 'before', e_cv.bias.shape: (8, 1, 1)
ic| 'after', e_cv.weight.shape: (8, 2, 4, 4)
ic| 'after', e_cv.bias.shape: (8, 1, 1)
ic| output_gen_torch.shape: torch.Size([2, 8, 4, 4])
ic| output_gen_jax.shape: (2, 8, 4, 4)
```

From that I can see that the PyTorch weights have the shape `t_cv.weight.shape: torch.Size([2, 8, 4, 4])` while the Jax weights have the shape `e_cv.weight.shape: (8, 2, 4, 4)`. Furthermore, you perform a VMAP at the bottom, which performs it over axis 0 by default. Given that, and that the weights have their first 2 dimensions swapped, perhaps that's an indicator? I'm not 100% sure.

**Comment by chMoussa** [comment]
@Artur-Galstyan Actually the first dimension of z can be considered a batch size (can be different from latent_space_dim). I also tried to permute the torch weights to match the equinox weights dimensions (either `t_cv.weight.detach().permute(1,0,2,3)` or `t_cv.weight.detach().permute(1,0,3,2)`), but the outputs still differs.

**Comment by chMoussa** [comment]
So it turns out that the weights of the kernels need flipping and then one needs to transpose on the first two axes. Here is the function that sets the parameters to obtain close results between Pytorch and Equinox:

```
import numpy as np
import jax.numpy as jnp
import jax.tree_util as jtu
import torch.nn as tnn
import equinox as eqx

def set_weights(t_cv: tnn.Module, e_cv: eqx.Module):
    t_w = t_cv.weight.detach().numpy().astype(np.float64)
    t_b = t_cv.bias.detach().numpy().astype(np.float64).reshape(e_cv.bias.shape)
    _, treedef = jtu.tree_flatten(e_cv)
    new_e_cv = jtu.tree_unflatten(treedef, [jnp.flip(t_w, axis=(2,3)).transpose(1,0,2,3), jnp.array(t_b)])
    return new_e_cv
```

The only thing is that the tolerance should be set to `1e-7`:
`assert jnp.isclose(output_gen_torch.numpy(), output_gen_jax).all()
`
